torch/testing/_internal/distributed/rpc/jit/rpc_test.py

```diff
@@ -1038,6 +1038,23 @@ def callback(fut):
 
         self.assertEqual(fut.wait(), torch.ones(n, n) + 1 + num_cbs)
 
+    @dist_init
+    def test_add_done_callback(self):
+        callback_called = None
+
+        def callback(_fut):
+            nonlocal callback_called
+            callback_called = fut.wait() * 2
+
+        future = rpc.rpc_async(
+            worker_name((self.rank + 1) % self.world_size),
+            script_fork_wait_udf,
+            args=(torch.ones(2),),
+        ).add_done_callback(callback)
+        self.assertFalse(callback_called)
+        self.assertEqual(future.wait(), torch.ones(2) * 2)
+        self.assertEqual(callback_called, torch.ones(2) * 4)
```

Comment:
If you let `future` be the return value of `rpc_async`, then the completion of `future` no longer guarantees the completion of the callback. You will need additional signals to indicate the callback is done.

Reply:
Right, thanks. I added `then` callbacks to those tests that should run after the `add_done_callback` which I now wait on. That should guarantee they run before the test finishes- it makes the test a little more convoluted, but given that we have separated tests devoted just to testing `then` I think that's fine.

Reply:
Nit: I think that this approach is fine, although it implicitly assumes an implementation detail, which is that the callbacks are called in the order that they are added (but the only time the order should matter to the user is when they explicitly chain callbacks with `.then(...).then(....)` etc. 

I think it's fine to assume this for now since the implementation is probably unlikely to change, although another way to test this could be to have the done callback set a `threading.Event` and wait on that.

Reply:
It looks like we do according to the docs here? https://github.com/pytorch/pytorch/blob/8a1e1004664e4cc8b1adbcef07b9fdae397e1787/torch/futures/__init__.py#L49

At least, it sounds like we make guarantees around the order that callbacks registered from `then` are called. And we have the same guarantee for callbacks registered from `add_done_callback`. I guess the fact that any "interleaving" of these calls will also be called in the same registration order is really an implementation detail, although that will probably be unlikely to change too. Either way thanks for offering the alternative.